Android/app/src/main/java/app/intra/DnsVpnAdapter.java

```diff
@@ -207,35 +327,46 @@ private String getProtocolErrorMessage(byte protocol) {
 
   @Override
   public void sendResult(DnsUdpQuery dnsUdpQuery, DnsTransaction transaction) {
-    // Construct a reply to the query's source port by switching the source and destination.
-    UdpPacket udpResponsePacket =
-        new UdpPacket(dnsUdpQuery.destPort, dnsUdpQuery.sourcePort, transaction.response);
-    byte[] rawUdpResponse = udpResponsePacket.getRawPacket();
-
-    IpPacket ipPacket;
-    if (dnsUdpQuery.sourceAddress instanceof Inet4Address) {
-      ipPacket = new Ipv4Packet(
-          UDP_PROTOCOL,
-          dnsUdpQuery.destAddress,
-          dnsUdpQuery.sourceAddress,
-          rawUdpResponse);
-    } else {
-      ipPacket = new Ipv6Packet(
-          UDP_PROTOCOL,
-          dnsUdpQuery.destAddress,
-          dnsUdpQuery.sourceAddress,
-          rawUdpResponse);
+    if (transaction.response != null) {
```

Comment:
Early return may be cleaner.

Reply:
We decided not to do this due to the final call to `recordTransaction`.